librz/util/sys.c

```diff
@@ -1380,3 +1383,248 @@ RZ_API void rz_sys_info_free(RSysInfo *si) {
 	free (si->machine);
 	free (si);
 }
+
+#if __UNIX__
+#if HAVE_PIPE2
+#elif defined(O_CLOEXEC)
+static RzThreadLock *sys_pipe_mutex;
+
+__attribute__ ((constructor)) static void sys_pipe_constructor(void) {
+	sys_pipe_mutex = rz_th_lock_new (true);
+}
+
+__attribute__ ((destructor)) static void sys_pipe_destructor(void) {
+	rz_th_lock_free (sys_pipe_mutex);
+}
+
+static bool set_close_on_exec(int fd) {
+	int flags = fcntl (fd, F_GETFD);
+	if (flags == -1) {
+		return false;
+	}
+	flags |= FD_CLOEXEC;
+	return fcntl (fd, F_SETFD, flags) != -1;
+}
+
+/**
+ * \brief Create a pipe and use O_CLOEXEC flag when \p close_on_exec is true
+ *
+ * If \p rz_sys functions are used to exec/system the new process, no race
+ * condition happen even on systems that don't support atomic creation of pipes
+ * with O_CLOEXEC.
+ */
+RZ_API int rz_sys_pipe(int pipefd[2], bool close_on_exec) {
+	int res = -1;
+	rz_th_lock_enter (sys_pipe_mutex);
+	if ((res = pipe (pipefd)) == -1) {
+		perror ("pipe");
+		goto err;
+	}
+	if (close_on_exec && (!set_close_on_exec (pipefd[0]) || !set_close_on_exec (pipefd[1]))) {
+		perror ("close-on-exec");
+		close (pipefd[0]);
+		close (pipefd[1]);
+		goto err;
+	}
+err:
+	rz_th_lock_leave (sys_pipe_mutex);
+	return res;
+}
+
+/**
+ * \brief Close a file descriptor previously created pipe \p rz_sys_pipe
+ */
+RZ_API int rz_sys_pipe_close(int fd) {
+	return close (fd);
+}
+
+RZ_API int rz_sys_execv(const char *pathname, char *const argv[]) {
+	rz_th_lock_enter (sys_pipe_mutex);
```

Comment:
Is this really necessary around the `execv` calls? I don't really see any races here.

Reply:
I think this is necessary to ensure an `execv` is not executed in between the `open` and the `fcntl`. If that would happen, you would have a pipe fd that leaks in the child process.

Reply:
Good point, maybe add a comment about this since it's a bit non-intuitive that locking around a single syscall is necessary.